MCH: introduce RoFRecord in decoder

aferrero2707 opened the pull request:
The MCH RoF are designed to group all SAMPA digits that have exactly the same time stamp.
Further grouping of the digits into time clusters is delegated to a later stage of the processing pipeline.

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/UserLogicElinkDecoder.h:
```diff
@@ -186,9 +186,15 @@ bool UserLogicElinkDecoder<CHARGESUM>::append10(uint10_t data10)
-          sendHBPacket();
-          transition(State::WaitingHeader);
-          result = true;
+          if (mSampaHeader.isHeartbeat()) {
```
this change is unrelated to this PR, isn't it ?

aferrero2707 replied:
Yes, indeed. I will remove the unrelated changes and prepare separate PR(s) for them.

aphecetche commented on Detectors/MUON/MCH/Workflow/src/DataDecoderSpec.cxx:
```diff
@@ -158,30 +174,44 @@ class DataDecoderTask
-    pc.outputs().adoptChunk(Output{"MCH", "DIGITS", 0}, digitsBuffer, digitsSize, freefct, nullptr);
-    pc.outputs().adoptChunk(Output{"MCH", "ORBITS", 0}, orbitsBuffer, orbitsSize, freefct, nullptr);
+    pc.outputs().adoptChunk(Output{header::gDataOriginMCH, "DIGITS", 0}, digitsBuffer, digitsSize, freefct, nullptr);
+    pc.outputs().adoptChunk(Output{header::gDataOriginMCH, "ROFS", 0}, rofsBuffer, rofsSize, freefct, nullptr);
```
other workflows expect the digit rofs spec to be named "DIGITROFS" (because there's also CLUSTEROFs, etc...)

aferrero2707 replied:
Ok, I will change the output spec.

aferrero2707 committed:
[MCH] preparing to remove time field from Digit structure

aferrero2707 committed:
[MCH] added ROFRecord output in raw decoder

The MCH RoF are designed to group all SAMPA digits that
have exactly the same time stamp.
Further grouping of the digits into time clusters is
delegated to a later stage of the processing pipeline.

The code also introduces a unit test for the RoF
finding algorithm.

aferrero2707 commented:
I just pushed some changes/simplifications to the ROF finder code. In particular:

- the time-based sorting of the digits is now applied on a vector containing the indexes of the digits in the input vector (while the old code was using a multiset of digit pointers). For simplicity the digits are not split into different orbits, and only a single global vector is sorted.
- the ROFFinder object does not need to be persistent, it is therefore instantiated each time the decoding  is completed. The ROFFinder constructor now takes a reference to the decoded digits vector and the value of the first orbit in the TF, and outputs a vector of ROFRecords and a vector of digit indexes sorted in time-ascending order. Two helper functions allow to store the ROF and sorted digits vectors into memory buffers that can then be passed to the DPL outputs.
- it is possible that some digits at the beginning of a given Time Frame are arriving late, and that their time stamp actually corresponds to the previous TF. At this stage those digits are kept and the corresponding ROFs are created with an orbit number smaller than the first orbit of the current TF. For the moment the proper handling of such out-of-TF digits is deferred to a later stage in the workflow...

pillot commented on Detectors/MUON/MCH/Raw/Decoder/src/ROFFinder.cxx:
```diff
@@ -0,0 +1,370 @@
+
+    return (dOrbit * BCINORBIT + dBC);
+  };
+
```
you could use the function InteractionRecord::differenceInBC(...) instead of this lambda.

aferrero2707 replied:
Done.

pillot commented on Detectors/MUON/MCH/Raw/Decoder/src/ROFFinder.cxx:
```diff
@@ -0,0 +1,370 @@
+  }
+
+  mRawDigitIdComp.digits = &mInputDigits;
+  std::sort(mOrderedDigits.begin(), mOrderedDigits.end(), mRawDigitIdComp);
```
I think you can use a lambda function instead of creating this structure

aferrero2707 replied:
Good suggestion. I  have implemented it and it will appear in the next commit.

pillot reviewed (commented):
The new logic looks fine to me. I just have have a couple of minor suggestions.

aferrero2707 committed:
[MCH] using lambda fuction for digits sorting

aferrero2707 committed:
[MCH] added ROFs in digits sink to text file

aphecetche reviewed (dismissed):
Just a remark (and certainly not a showstopper for this PR), but please note that the rofs output is in principle already included in the `o2-mch-digits-file-writer-workflow` (see https://github.com/AliceO2Group/AliceO2/tree/dev/Detectors/MUON/MCH/DevIO/Digits). The idea being to remove the digits-sink and other related  "debug-only" from the workflow directory (and keep there only the "production" code) (at least once we confirm that the ones in DevIO library are working ok).

aferrero2707 commented:
The CI system seems to have gone bananas again... @ktf any idea?

aferrero2707 committed:
[MCH] fixed resizing of digits buffer when creating dummy ROFs

aphecetche commented:
@ktf @davidrohr @TimoWilken what's the situation with the CI ?

ktf commented:
See https://alice.its.cern.ch/jira/browse/O2-2356 for something which is at least partially related.

ktf commented:
Simulation issue unrelated. Merging.
